### ISSUE TITLE
Type of existing variable is not being specified when statement in assignment returns Awaitable 

### ISSUE THREAD

**Issue by kammala** [python/mypy#10516]
**Bug Report**
When using variable with predefined union type in assignment statement, its type is not being corrected according to return type of assigned statement.

**To Reproduce**
The simplest synthetic(but still close to real life) example would be:
```python
import asyncio

from typing import Optional


def g() -> int:
    return 1


async def f(x: Optional[int] = None) -> int:
    loop = asyncio.get_running_loop()

    if x is not None:
        return x
    x = await loop.run_in_executor(None, g)
    return x
```
It is also reproducible on synthetic functions(like `def returns_awaitable(var: T) -> Awaitable[T]` instead of `run_in_executor`), so I don't think this bug relates to `run_in_executor` specifically.

**Expected Behavior**
I would expect `reveal_type(x)` return `int`(according to [typeshed](https://github.com/python/typeshed/blob/master/stdlib/asyncio/base_events.pyi#L80) `run_in_executor` returns `Future[int]`).

**Actual Behavior**
It seems that it is being resolved to`Optional[int]` – original type of `x`.
```
example.py:17: error: Incompatible return value type (got "Optional[int]", expected "int")  [return-value]
```

Weird enough that splitting assignment into 2 parts works as expected, this snippet doesn't generate an error:
```python
...
fut = loop.run_in_executor(None, g)
x = await fut
return x
```

**Environment**

- Mypy version used: 0.812
- Mypy command-line flags: `--show-error-codes --skip-cache-mtime-checks`
- Mypy configuration options from `mypy.ini` (and other config files):
```
[mypy]
disallow_any_decorated = True
disallow_incomplete_defs = True
disallow_untyped_calls = True
disallow_untyped_defs = True
ignore_missing_imports = True
strict_equality = True
warn_unreachable = True
warn_unused_configs = True
warn_unused_ignores = True
```
- Python version used: 3.9.2
- Operating system and version: macOS 11.3.1

**Comment by erictraut** [comment]
This bug appears to have been fixed. I can't repro it with the latest version of mypy (1.5).
